File: app/src/main/java/com/example/tetrisapp/util/ConnectionUtil.java
package com.example.tetrisapp.util;

import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.net.NetworkRequest;

import androidx.annotation.NonNull;

import java.net.InetAddress;

public class ConnectionUtil {
    ConnectivityManager connectivityManager;
    NetworkRequest networkRequest = new NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
            .build();
    Callback onAvailable;
    Callback onLost;

    public ConnectionUtil(ConnectivityManager connectivityManager) {
        this.connectivityManager = connectivityManager;
    }

    public boolean isNetworkConnected() {
        NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
        if (networkInfo == null) return false;

        return networkInfo.isConnected();
    }

    public boolean isInternetAvailable() {
        try {
            InetAddress ipAddress = InetAddress.getByName("google.com");
            return !ipAddress.equals("");
        } catch (Exception e) {
            return false;
        }
    }

    public void checkInternetConnection() {
        ConnectivityManager.NetworkCallback networkCallback = new ConnectivityManager.NetworkCallback() {
            @Override
            public void onAvailable(@NonNull Network network) {
                super.onAvailable(network);
                if (isInternetAvailable()) {
                    onAvailable.call();
                } else {
                    onLost.call();
                }
            }

            @Override
            public void onLost(@NonNull Network network) {
                super.onLost(network);
                onLost.call();
            }
        };

        connectivityManager.requestNetwork(networkRequest, networkCallback);
    }

    public void setOnAvailable(Callback onAvailable) {
        this.onAvailable = onAvailable;
    }

    public void setOnLost(Callback onLost) {
        this.onLost = onLost;
    }

    public interface Callback {
        void call();
    }
}

File: app/src/main/java/com/example/tetrisapp/util/ConnectionUtil.kt
package com.example.tetrisapp.util

import android.content.Context
import android.net.ConnectivityManager
import android.net.ConnectivityManager.NetworkCallback
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.util.Log

class ConnectionUtil(ctx: Context) {
    private val connectivityManager: ConnectivityManager
    private val networkRequest: NetworkRequest = NetworkRequest.Builder()
        .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
        .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
        .build()

    var onAvailableCallback: (Network) -> Unit = {}
    var onLostCallback: (Network) -> Unit = {}

    init {
        connectivityManager = ctx.getSystemService(ConnectivityManager::class.java) as ConnectivityManager
    }

    fun checkInternetConnection() {
        val networkCallback: NetworkCallback = object : NetworkCallback() {
            override fun onAvailable(network: Network) {
                super.onAvailable(network)
                onAvailableCallback(network)
            }

            override fun onLost(network: Network) {
                super.onLost(network)
                onLostCallback(network)
            }
        }
        connectivityManager.requestNetwork(networkRequest, networkCallback)
    }

    fun isOnline(): Boolean {
        val capabilities =
            connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
        if (capabilities != null) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_CELLULAR")
                return true
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_WIFI")
                return true
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_ETHERNET")
                return true
            }
        }
        return false
    }

    companion object {
        const val TAG = "ConnectionUtil"
    }
}